### DISCUSSION TITLE
Math formulas

### DISCUSSION THREAD

**Discussion by horstfortytwo** [knowledgefutures/pubpub#1869] [Ideas]
I am using pubpub to write a book (course/lecture notes) on mathematical logic. I am loving pubpub, but: The text I write contains on average one or two math formulas per sentence and the WYSIWYG editor of pubpub drives me crazy in this respect. 

To make this more precise, here is what I have to do to insert a formula (even a simple x):
1. I move my mouse/cursor to the formula button and click on it. A default formula is inserted into the text.
2. I click on the standard formula. A pop-up window opens, displaying the default formula.
3. I click on the pop-up window (to gain focus).
4. I erase the default formula.
5. I click on the update button.

This is so laborious that I now do the following as a work around. I type the entire text in [CodiMD](https://hackmd.okfn.de/), which is a mark down editor with math support and regularly remove the entire content from pubpub and import the md file from CodiMD.

Here is what would make my work easier and is probably easy to implement.
- A keyboard short-cut that opens a pop-up window for directly editing math, for instance, CTRL-M for an inline formula und CTRL-SHIFT-M for a displayed formula.
- The pop-up window is empty, so no default formula.
- The focus is directly in the edit area of the pop-up area. 
- A keyboard short-cut for saving (updating) the formula.
The point here is that this can all be done using the keyboard. No mouse interaction necessary, only a few additional key strokes for entrering a math formula. In fact, this would be as fast and convenient as surrounding math formulas by $ or $$, as is usual in other contexts.

Maybe, this is the wrong place to discuss this issue. I would be very grateful for any advise where to post it instead.

**Comment by gabestein** [comment]
Hi – this is the right place to post. Thanks very much for using PubPub, and for the feedback on the difficulty writing quickly with math. It makes sense to me, especially as a power-user of keyboard shortcuts on other tools. We'll discuss on our eng call next week and see what's possible.

**Reply by horstfortytwo** [reply]
This is really good to hear! Thank you very much looking into this!👍

**Reply by gabestein** [reply]
Thanks again for the suggestion. We had a nice discussion about this today, and are going to do a research spike to look at ways we can make editing with math easier in our next sprint (https://github.com/pubpub/pubpub/issues/1917).

**Reply by horstfortytwo** [reply]
I am looking forward to the results of your research! Thanks a lot! 🙏

**Reply by gabestein** [reply]
Just wanted to let you know that this is almost ready. We've got the basics working, but have a few UIs to work out to replace some of the settings that were formerly in the LaTeX writing panel.

**Comment by horstfortytwo** [comment]
Here is a good example for what I am trying to do: https://logic.pubpub.org/pub/o3hrnyf8/release/3 

Many, many math formulas ...

**Comment by gabestein** [comment]
Heads up: we launched the new math editor into production today. There are a few things we need to address (mostly css and making them referenceable), but it should be a much-improved experience. Thanks again for the suggestion!

**Reply by horstfortytwo** [reply]
That sounds great. I played around with the new feature ... very nice. BUT: When I have a text consisting of x^2 only, the md download fails. Uploading md files with math does not work either. That was no problem in earlier versions. I am sorry I have to say that.

![image](https://user-images.githubusercontent.com/101699838/179912851-a76d6a8b-f690-4a53-bb95-c6b57c99e9a8.png)

**Reply by horstfortytwo** [reply]
And here are more details about the upload:

When the md file contains `$x^2$`, an error is produced.
When the md file contains `\$x^2\$`, no error is produced, but there is no math mode.
When the md file contains `\\$x^2\\$`, an error is produced.
When the md file contains `$$x^2$$`, an error is produced.

**Reply by gabestein** [reply]
Ah, thanks for catching that. We’ll need to wire the new math nodes up to pandoc. Shouldn’t be a huge effort.

**Reply by horstfortytwo** [reply]
Whatever that means! ;-) But thanks anyway!

**Reply by gabestein** [reply]
Ok, I think this is fixed.

**Reply by horstfortytwo** [reply]
Thanks very much. It seems to me that the upload works smoother than before working on the math module. Thanks again!
